server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/Email.scala

```diff
@@ -109,6 +224,198 @@ case class EmailBody(bodyStructure: EmailBodyPart,
+    val messageId: MessageId = message._1
+    val mailboxIds: MailboxIds = MailboxIds(message._2
+      .map(_.getMailboxId)
+      .toList)
+
+    for {
+      firstMessage <- message._2
+        .headOption
+        .map(Success(_))
+        .getOrElse(Failure(new IllegalArgumentException("No message supplied")))
+      blobId <- BlobId.of(messageId)
+    } yield {
+      EmailMetadataView(metadata = EmailMetadata(
+        id = messageId,
+        blobId = blobId,
+        threadId = ThreadId(messageId.serialize),
+        mailboxIds = mailboxIds,
+        receivedAt = UTCDate.from(firstMessage.getInternalDate, zoneIdProvider.get()),
+        size = sanitizeSize(firstMessage.getSize)))
+    }
+  }
+}
+
+private class EmailHeaderViewFactory @Inject()(zoneIdProvider: ZoneIdProvider) extends EmailViewFactory {
+  override def toEmail(request: EmailGetRequest)(message: (MessageId, Seq[MessageResult])): Try[EmailView] = {
+    val defaultMessageBuilder = new DefaultMessageBuilder
+    defaultMessageBuilder.setMimeEntityConfig(MimeConfig.PERMISSIVE)
+    defaultMessageBuilder.setDecodeMonitor(DecodeMonitor.SILENT)
+    val messageId: MessageId = message._1
+    val mailboxIds: MailboxIds = MailboxIds(message._2
+      .map(_.getMailboxId)
+      .toList)
+
+    for {
+      firstMessage <- message._2
+        .headOption
+        .map(Success(_))
+        .getOrElse(Failure(new IllegalArgumentException("No message supplied")))
+      mime4JMessage <- Try(defaultMessageBuilder.parseMessage(firstMessage.getFullContent.getInputStream))
+      blobId <- BlobId.of(messageId)
+    } yield {
+      EmailHeaderView(metadata = EmailMetadata(
+        id = messageId,
+        blobId = blobId,
+        threadId = ThreadId(messageId.serialize),
+        mailboxIds = mailboxIds,
+        receivedAt = UTCDate.from(firstMessage.getInternalDate, zoneIdProvider.get()),
+        size = sanitizeSize(firstMessage.getSize)),
+        header = EmailHeaders.from(zoneIdProvider.get())(mime4JMessage))
+    }
+  }
+}
+
+private class EmailFullViewFactory @Inject()(zoneIdProvider: ZoneIdProvider, previewFactory: Preview.Factory) extends EmailViewFactory {
+  override def toEmail(request: EmailGetRequest)(message: (MessageId, Seq[MessageResult])): Try[EmailView] = {
+    val defaultMessageBuilder = new DefaultMessageBuilder
+    defaultMessageBuilder.setMimeEntityConfig(MimeConfig.PERMISSIVE)
+    defaultMessageBuilder.setDecodeMonitor(DecodeMonitor.SILENT)
+    val messageId: MessageId = message._1
+    val mailboxIds: MailboxIds = MailboxIds(message._2
+      .map(_.getMailboxId)
+      .toList)
+
+    for {
+      firstMessage <- message._2
+        .headOption
+        .map(Success(_))
+        .getOrElse(Failure(new IllegalArgumentException("No message supplied")))
+      mime4JMessage <- Try(defaultMessageBuilder.parseMessage(firstMessage.getFullContent.getInputStream))
```

Comment:
Is it our responsability to close the stream here ?

Reply:
Likely. There should be a cool monad for this?